### PR TITLE
Added SecretFileMode for specifying mounted secrets file permissions.

### PR DESCRIPTION
**Pull request by soltysh**
This came as a result of discussion in #4789. 
@pmorie @jimmidyson PTAL.

### CODE REVIEW

**Comment by pmorie** [comment]
@soltysh This needs another case in the secrets e2e

**Comment by pmorie** [comment]
@soltysh I think that new case case be, create different files with different modes and use the mount tester binary to check them.  You can only check one file's mode with the current mount tester so it will probably 1 case per non-standard mode

**Comment by erictune** [comment]
Unless there is a good reason to rush this, I'd like to put a hold on this
till I (we?) understand #7925 better.
For one thing, I'm not sure whether the permissions belong on the Secret or
the SecretVolumeSource.

**Comment by pmorie** [comment]
@erictune 

> For one thing, I'm not sure whether the permissions belong on the Secret or
the SecretVolumeSource.

I'm actually thinking now that #7925 is what we ought to be talking about.  Once we have security and volumes sorted there shouldn't be any need to control permissions -- secrets in volumes should always be 0400, owned by the uid in the pod's security context, and have the selinux context set correctly.

**Comment by soltysh** [comment]
That was my first take in resolving #4789, just to set by default 0400 on mounted secrets. Still that kind of action generates risk that since we mount those as root, only root will have access to it. I then started chatting with @pmorie about that and we came up with this solution. I'll then wait for #7925 and its output.

File: pkg/api/types.go [hunk 1]
```diff
@@ -450,6 +452,16 @@ type GitRepoVolumeSource struct {
 type SecretVolumeSource struct {
 	// Name of the secret in the pod's namespace to use
 	SecretName string `json:"secretName"`
+	// Modes describes access permissions to be applied per Secret's data key while mounting
+	Modes []SecretFileMode `json:"modes"`
```
Comment (bgrant0607) [inline pkg/api/types.go]: Does this really need to be per-key?

Comment (soltysh) [inline pkg/api/types.go]: You may want to mount each data's value with different permissions.

Comment (bgrant0607) [inline pkg/api/types.go]: Create multiple secrets.

Comment (soltysh) [inline pkg/api/types.go]: If we provide users with option to put multiple data we should allow them to define different permissions as well. On the other hand I can think of that as a grouping factor, where you'd group kind of "similar" data into one secret. Before doing any changes I'll wait for the outcome from #7925 @erictune mentioned.

**Comment by erictune** [comment]
Not sure if this is still active?  If it isn't, it can be closed.

At any rate, I don't think there will be time to complete review of this prior to the pre-v1 code freeze.

**Comment by soltysh** [comment]
I'm closing this until this gets sorted out, but everything seems that it'll be done differently.

**Pull request closed without merging**